File: BSTIterator.java
// Time Complexity : hasnext : O(1), next : O(h) in worst case but amortixed is O(1) averagely
// as for n/2 level (leafs) - O(1) , for one level up (n/4) O(1) - 75% of them complete in O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/* Approach:
We use our own stack as we have to control the recursion.
We go towards the left, when left is done, we process the root, and go towards the right
*/
class BSTIterator {
    Stack<TreeNode> st;
    public BSTIterator(TreeNode root) {
        this.st = new Stack<>();
        helper(root);        
    }
    private void helper(TreeNode root){
        while(root!=null){
            st.push(root);
            root = root.left;   
        }
    }
    
    public int next() { //O(h) ~ O(1)
        TreeNode r = st.pop();
        helper(r.right);
        return r.val;
        
    }
    
    public boolean hasNext() { //O(1)
        return (!st.isEmpty());
    }

}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: bstiterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    '''
    SC : O(H) 
    
    DFS Next 
    TC : O(H) // traverse the left until there is no left 
    SC : O(1) 
    
    Hasnext 
    TC : O(1)
    SC : O(1)
    
    Approach : 
    dfs() : traverse to the left most and store all the elements
    next() : pop the element from stack and expplore the right tree
    hasnext() : if any element left in stack

    '''
    
    def __init__(self, root: TreeNode):
        self.st = []
        self.dfs(root)

    def next(self) -> int:
        ans = self.st.pop()
        self.dfs(ans.right)
        return ans.val  

    def hasNext(self) -> bool:
        return True if self.st else False
    
    def dfs(self,root):
        while root:
            self.st.append(root)
            root = root.left



# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()